### DISCUSSION TITLE
Sync 2 or more ESP32 at milliseconds level

### DISCUSSION THREAD

**Discussion by alessandrv** [micropython/micropython#14173] [Hardware & Peripherals]
Hello, as the titles says I'm trying to figure out a way to sync 2 or more boards with perfect timing with at least 3 digit milliseconds. 
I'm not sure which approach would be the best for this. 
Trying to sync epoch time between 2 identical time grounding the pin at the same time gives random milliseconds from the ticks_ms() function where one has a bigger number first and maybe after 50ms the other one has the larger one for no reason, of course this doesn't fit my need. 
What do you guys suggest? I read about gps modules but can I trust them to be synced perfectly? I need this setup to be portable, so no home network, the best that I could have with me is phone hotspot but I think syncing the time with wifi would cause other delays issues. Do any rtc modules keep track of milliseconds where grounding a pin at the same time would grant me them to be synced?
Thanks in advance.

**Comment by glenn20** [comment]
Are you using Pin.irq handlers to sample the grounding of the pin or are you polling in a tight loop? The former will have higher and more unpredictable latency than the latter.

**Reply by alessandrv** [reply]
Just using a while True to print the ms when button value Is true to see if button Is pressed, so that should be the latter i think

**Reply by plus10db** [reply]
This may help: https://www.reddit.com/r/embedded/comments/1pbp0az/esp32_s3_submicrosecond_time_sync_and_disciplined/

**Comment by CxZMoE** [comment]
maybe we can use one ESP32 as local time synchronization server, itself sync time from internet ntp server. then connect other esp devices to that ESP32 server like a mesh and receiving synchronized time info from it periodly by broadcast?

**Reply by alessandrv** [reply]
That would make sense, but comunicating the time to other devices via network Will for sure not be perfectly synced between 2 devices. To Sync them its fine to me to use a wire, the issue Is that ms mesured by the ESP are not the same

**Reply by karfas** [reply]
[NTP](https://en.wikipedia.org/wiki/Network_Time_Protocol) exists since ~40 years, accounts for transmission delays and is able to synchronize up to some milliseconds time difference.
As far as I know no "full" NTP client or server exists for micropython, but a SNTP client comes with ESP-IDF.

**Reply by alessandrv** [reply]
Having 2 boards making a request to and ntp server wont have different ms because that would depend on network delay the time they take to get the reply with actual time might be After 100ms already went thourgh and maybe the other One took 10ms to get a reply

**Reply by karfas** [reply]
It's no rocket science to measure the time between send and receive. Most likely, a NTP server also notes in the response when the request had been received. 
Armed with these informations, it's not too complicate to correct the time received...

**Reply by alessandrv** [reply]
Sorry, i meant that giving those data to the others esp32 which are located at more than 500 meters from each other while using an extra esp32 to gather the ntp time would give other delay to the request because I would need to send those data to each other with rf module that would bring other latencies. And I would need this to happen frequently if my milliseconds timer goes off after less than a minute, and all of this is for a sport timer (my bad I didnt mention it) so the timing must be precise between devices all the time. I was thinking about using a gps module with a pps pin to reset the milli timer everytime I get the pulse, those should all sync together and that would happen every second

**Reply by robert-hh** [reply]
Following the discussions here, It seems that a GPS receiver of the same model at each ESP32 seems the best option. The PPS signal can be used with Pin IRQ top capture the last value of ticks_ms() and then you'd have a 1 second window to get the real time/date from GPS. Then you have  a value pair of ticks_ms() and date/time at both devices. The latency for a Pin.IRQ signal to calling a handler according to my tests is about 50µs if the code to be executed is cached. So it may be useful to let that PPS catch run a few times with no other activity at the ESP32 to ensure caching. If the code to be executed on the PPS pulse is not cached, the latency will be much larger. I a quick test right now I have seen 17 ms, but it can be much more.

**Reply by alessandrv** [reply]
Hello, I'm new to micropython, what you mean by caching the code? Do you have any documentation or repo here on github that I could check about that? I was planning on resetting the timer when the pin had power of course, then get the nmea message to get hour/min/sec.millisecondstimer, I would need to get the milli timer when a PIN is HIGH and that to be the precise number but have no idea on what you mean by caching here. Thanks in advance

**Reply by robert-hh** [reply]
The ESP32 has like many small controllers a serial flash memory and, if present, a serial RAM. So if code is to be executed, it has to be fetched though a serial SPI connection, which is comparable slow. To make that a  little bit faster, a cache is placed between CPU and flash, which speeds up the code fetch for pages located in the cache. That is a hardware mechanism and fully transparent to the code. But if the requested memory area is not in the, it has to be read from the flash memory. Similar for the so called SPIRAM.
Besides that, the IRQ at the ESP32 port is not a hard one.  It is scheduled as Python task, which can be delayed by other internal tasks.
That said, you should take care that for the moment of synchronization the device is mostly idle.

**Reply by alessandrv** [reply]
well the device may or may not be idle during the sync because I was planning on syncing every pps and the function that takes time could start at the same of the sync, wouldn't that bring possible delays that caching would remove? Maybe I'm understanding this wrong. At the moment when I use a button on a pin to check at what value the ms timer is often one is 1ms above, and maybe if I check later they are back in sync, like if the checking of the pressing of the button would start a ms after that on one of the boards (they switch: one time board 1 is above, maybe then they have the same ms value and maybe later the second is above), looks like a faster resolution of the checking if the button is pressed would make sense to fix this inconsistency that I don't understand? Or do you have any tip about this?

**Reply by alessandrv** [reply]
Like at the moment I just have a while True that if they button is pressed prints the ms timer value and then resets it. Sometime if I'm holding the button the print gives me 1 as output instead of 0, so there's some random delay I do not understand

**Reply by robert-hh** [reply]
That looks just like the +/- 1 display error of a digital value of an analog quantity time.

**Reply by alessandrv** [reply]
Not sure what I should do to avoid that then

**Reply by robert-hh** [reply]
So you reset the ms timer by e.g. remembering the actual time.ticks_ms() value and then you fetch the actual timer value (time.ticks_ms() - remembered_value) and print that? If you call time.ticks_ms() twice in a row it may have different values.
Edit: Just run:
```
import time
for _ in range(1000):
    if time.ticks_ms() - time.ticks_ms():
        print(_)
```

**Reply by alessandrv** [reply]
At the moment I have a timer class like this: 
```
class MillisecondTimer:
    def __init__(self):
        self.start_time = utime.ticks_ms()

    def elapsed_ms(self):
        return utime.ticks_diff(utime.ticks_ms(), self.start_time)
```

Running your code snipper I do print some results which I think is concerning as I should not see any (logically speaking). But sometime I get 5 prints, sometimes 1 but never zero.

**Reply by robert-hh** [reply]
> as I should not see any (logically speaking)

You should expect to see some. The numbers are expected to vary. The ticks timer is incremented every ms, and ticks_ms() just returns the actual value. If the first call of ticks_ms() is at the very end of one 1ms time slot, and the next call falls into the next or later time slot, then there is a difference between the two values. You cannot avoid that, unless you try to sync the calls of ticks_ms() with the internal ticks timer. And you should always expect that +/-1 error of the underlying time unit.

Edit: If you use ticks_us() in my above example and print the time difference, you get 5µs as the time difference between the calls, and thus with ms granularity in the test you can expect in the average 5 prints out of 1000 attempts.

**Reply by alessandrv** [reply]
Thank you, makes sense, but having 2 boards and checking the time at the same moment still gives me 2 different results which is not acceptable in sports timing unfortunately so I don't know what to do to avoid this still, 1ms doesn't sound like something huge but i cant provide something that gives wrong values even if only 1 ms

**Reply by robert-hh** [reply]
Maybe you can use ticks_us() as the base timer and round to ms(). With ticks_us() the intrinsic error of sampling is lower. But ticks_us() rolls over after 17 Minutes. 

b.t.w.: ticks_us() calls internally esp_timer_get_time() returning a 64 bit quantity, which should not overflow in reasonable time (~584544 years). So you could consider adding an interface for that value, if the overflow cannot be handled differently.

**Reply by alessandrv** [reply]
From my tests using ticks_us() and dividing to get ms is working better (I wonder why, thought that ticks_ms() was just doing that), still they 2 boards are off as soon as the timer start and there's always the risk of one getting to the next ms before the other, but it looks better, I don't think there's any better way to do this? I find it strange there's no external module/rtc to keep track of ms reliably

**Reply by robert-hh** [reply]
Actually ticks_ms() returns `esp_timer_get_time() /1000`, and ticks_us() return `esp_timer_get_time()`. So there should be no major difference.

> I find it strange there's no external module/rtc to keep track of ms reliably

What do you expect? The sampling problem will not disappear, the problem of delayed ISR will also stay. And if you work with buttons and switches, you'll have issues with bouncing.

**Comment by sosi-deadeye** [comment]
Maybe use for both ESP32 one **32_768 Hz** square wave generator as the clock source. But then you have to change the settings in [sdkconfig](https://github.com/micropython/micropython/issues/3764). You also need to know the right **Pin** on your ESP32-Dev-Board. If the 32kHz generator has a drift, both ESP32 should have the same drift in time.

**Comment by jonnor** [comment]
Are you trying to sync the RTC time, the internal clocks, or just synchronize an action (like sampling a sensor)? The available strategies will be quite different...

**Reply by alessandrv** [reply]
At the moment for my tests  I was just trying to sync ms between 2 devices. When printing the ms value % 1000 + the RTC time I was getting inconsistent seconds with the ms but that makes sense because the rtc is not looking at the same ms I'm getting but that's not the issue. I need 2 devices to have the same ms mainly, syncing the seconds shouldn't be too hard. I was planning on using a GPS Module with a pps pin so that every impulse I get I can reset my timer so every second I have a new Timer that should say consistent during a single second, does it make sense? Also the gps would give me the right second from its nmea message. In my op I forgot to mention that what I'm doing is a sport timer, so I need to measure the time it takes to get from point A to point B and so they must have their time synced perfectly. So to answer your question I just need to have a local time that's the same on 2+ boards, no need it to be the real time just them to match, so any solution would be great that sync them like this

**Reply by NikolaPU** [reply]
Hi Alessandrv, 
did you manage to sync 2 ESP devices ?
Did you use GPS to sync internal RTC ?

**Reply by alessandrv** [reply]
> Hi Alessandrv, did you manage to sync 2 ESP devices ? Did you use GPS to sync internal RTC ?

I ended up using an external ocxo to have a lot of stability. Use the pcnt of the esp to count the pulses to get the time. To sync just start the pcnt at the same time with an interrupt on both caused by a pin grounded at the same time. Not the most precise solution but if using an ocxo at 10mhz with that high resolution having them off by like 100 pulses is like them being off about 10 microseconds, so would do for now. My solution is not this straight forward, I do some other stuff to reduce the pulse difference between devices but this is the general concept I'm trying at the moment.
